README.md
# AgriTech-USGS-LIDAR-Challenge

- Free software: [MIT license](https://github.com/Bethelsis/AgriTech-USGS-LIDAR-Challenge/blob/main/LICENSE)
- This project creates a python module that domain experts and data scientists can use to fetch, visualise, and transform publicly available satellite and LIDAR data,simplifying LIDAR point cloud processing and rapid prototyping.

## Data

- LIDAR high definition elevation data - https://www.usgs.gov/core-science-systems/ngp/3dep

## Installation

- The package is not available PyPi yet. You can download the github repository and use it.

## Code

- visualize.ipynb- contains the data loading and visualizations.
- scripts folder -contains the data loading, preprocessing and vizualization functions. 

## Dependencies

- To install the necessary dependencies, execute the command $ pip install -r requirements.txt




CHANGELOG.txt
Change Log
==========

0.0.1
----------
-First version





scripts/test_fetch_data.py
from fetch_lidar_data import fetch_lidar_data
from shapely.geometry import Polygon, Point

MINX, MINY, MAXX, MAXY = [-93.756155, 41.918015, -93.756055, 41.918115]
polygon = Polygon(((MINX, MINY), (MINX, MAXY), (MAXX, MAXY), (MAXX, MINY), (MINX, MINY)))
bound = '([-10425171.940, -10423171.940], [5164494.710, 5166494.710])'
region = 'IA_FullState'
fetch_data = FetchData(bound, region)
fetch_data.run_pipeline()


tests/test_fetch_lidar_data.py
import unittest
import sys
sys.path.append("..")
from scripts.region_selector import dataset_path
from scripts.region_selector import SelectName

class Test_region_selector(unittest.TestCase):

    def test_SelectName(self):
        self.assertEqual(print(SelectName("IA_FullState")), "IA_FullState")

    def test_dataset_path(self):
        self.assertEqual(print(dataset_path("IA_FullState"), "https://s3-us-west-2.amazonaws.com/usgs-lidar-public/IA_FullState/ept.json"))


if(__name__ == '__main__'):
    unittest.main()


scripts/region_selector.py
import io

def SelectName(region:str):
    name_ls = []
    names_list =  io.open('../data/filename.txt', encoding='UTF-8').read().strip().split('\n')
    
    if region in names_list or region == 'IA_FullState':
        return region

    if region == 'all':
        return names_list

    else:
        for words in names_list:
            words_ls = words.split('_')
            if region in words_ls:
                name_ls.append(words)
            else: continue
        if name_ls == []:
            return None
        else: return name_ls

def dataset_path(region:str):
    name_list = SelectName(region)

    if type(name_list) == str:
        full_dataset_path = f'https://s3-us-west-2.amazonaws.com/usgs-lidar-public/{name_list}/ept.json'
        tif_path = name_list+'.tif'
        laz_path = name_list+'.laz'
        return full_dataset_path, tif_path, laz_path

    if type(name_list) == None:
        full_dataset_path = f'https://s3-us-west-2.amazonaws.com/usgs-lidar-public/{name_list}/ept.json'
        tif_path = f'{name_list}.tif'
        laz_path = f'{name_list}.laz'
        return full_dataset_path, tif_path, laz_path
    
    if type(name_list) == list:
        full_dataset_path_list = []
        tif_path_list = []
        laz_path_list = []
        for i in name_list:
            path = 'https://s3-us-west-2.amazonaws.com/usgs-lidar-public/'+i+'/ept.json'
            tif = i+'.tif'
            laz = i+'.laz'
            tif_path_list.append(tif)
            laz_path_list.append(laz)
            full_dataset_path_list.append(path)
        return full_dataset_path_list, tif_path_list, laz_path_list



scripts/fetch_lidar_data.py
import json
import pdal
import geopandas as gpd
from shapely.geometry import Polygon, Point
from region_selector import dataset_path
from edit_pipeline import edit_pipeline
from logger import Logger
logger = Logger()
logger = logger.get_logger('FetchData')
class fetch_lidar_data():
    def __init__(self, polygon, region:str, input_epsg=3857, output_epsg=4326):

        self.polygon = polygon
        self.region = region
        self.input_epsg = input_epsg
        self.output_epsg = output_epsg

    def get_polygon_boundaries(self, polygon: Polygon):
        polygon_df = gpd.GeoDataFrame([polygon], columns=['geometry'])

        polygon_df.set_crs(epsg=self.output_epsg, inplace=True)
        polygon_df['geometry'] = polygon_df['geometry'].to_crs(epsg=self.input_epsg)
        minx, miny, maxx, maxy = polygon_df['geometry'][0].bounds

        polygon_input = 'POLYGON(('
        xcords, ycords = polygon_df['geometry'][0].exterior.coords.xy
        for x, y in zip(list(xcords), list(ycords)):
            polygon_input += f'{x} {y}, '
        polygon_input = polygon_input[:-2]
        polygon_input += '))'
        return f"({[minx, maxx]},{[miny,maxy]})", polygon_input


    def create_pipeline(self):
        bound, polygon_input = self.get_polygon_boundaries(self.polygon)
        full_dataset_path, tif_path, laz_path = dataset_path(self.region)
        if type(full_dataset_path) == str:
            edited_json = edit_pipeline(full_dataset_path, laz_path, tif_path, self.output_epsg, bound, polygon_input)
            pipeline = pdal.Pipeline(json.dumps(edited_json))
            return pipeline
        else:
            pipeline_list = []
            for i in range(len(tif_path)):
                edit_json = edit_pipeline(full_dataset_path[i], laz_path[i], tif_path[i], bound)
                pipelines = pdal.Pipeline(json.dumps(edit_json))
                pipeline_list.append(pipelines)
            return pipeline_list



    def run_pipeline(self):
        pipelines = self.create_pipeline()
        if type(pipelines)==list:
            metadata_list = []
            log_list = []
            for pipeline in pipelines:
                try:
                    pipeline.execute()
                    metadata = pipeline.metadata
                    log = pipeline.log
                    metadata_list.append(metadata)
                    log_list.append(log)
                    return pipeline.arrays
                except RuntimeError as e:
                    print(e)
                    continue

        else:
            pipelines.execute()
            metadata = pipelines.metadata
            log = pipelines.log
            return pipelines.arrays
    def get_elevation(self, array_data):
        if array_data:

            for i in array_data:
                geometry_points = [Point(x, y) for x, y in zip(i["X"], i["Y"])]
                elevations = i["Z"]
                df = gpd.GeoDataFrame(columns=["elevation", "geometry"])
                df['elevation'] = elevations
                df['geometry'] = geometry_points
                df = df.set_geometry("geometry")
                df.set_crs(epsg=26915, inplace=True)

            return df            

if(__name__ == '__main__'):
    MINX, MINY, MAXX, MAXY = [-93.756155, 41.918015, -93.756055, 41.918115]
    polygon = Polygon(((MINX, MINY), (MINX, MAXY), (MAXX, MAXY), (MAXX, MINY), (MINX, MINY)))
    region = "IA_FullState"
    data_fetcher = FetchData(polygon, region)
    pipeline_list = data_fetcher.create_pipeline()
    data=data_fetcher.run_pipeline()
    df = data_fetcher.get_elevation(data)
    print(df.info())
    print(df)



fetch_data.py
import pdal
import json
import geopandas as gpd
from shapely.geometry import box, Point, Polygon
class Fetch_data:
    def __init__(self,bound,region:str) -> None:
        self.bound = bound
        self.region = region
    def read_Json(self) -> dict:
        try:
            with open("fetch_data.json", 'r') as json_file:
                dict_obj = json.load(json_file)
            return dict_obj
        except FileNotFoundError as e:
            print('FILE NOT FOUND')
    def polygon_boundaries(self, polygon: Polygon):
        polygon_df = gpd.GeoDataFrame([polygon], columns=['geometry'])

        polygon_df.set_crs(26915, inplace=True)
        polygon_df['geometry'] = polygon_df['geometry'].to_crs(3857)
        minx, miny, maxx, maxy = polygon_df['geometry'][0].bounds
        polygon_input = 'POLYGON(('
        xcord, ycord = polygon_df['geometry'][0].exterior.coords.xy
        for x, y in zip(list(xcord), list(ycord)):
            polygon_input += f'{x} {y}, '
        polygon_input = polygon_input[:-2]
        polygon_input += '))'
        return polygon_input

    def getPipeline(self, polygon: Polygon):
        fetch_json = self.read_Json()
        polygon_input = self.polygon_boundaries(polygon)
        full_dataset_path = f"https://s3-us-west-2.amazonaws.com/usgs-lidar-public/{region}/ept.json"
        #print(full_dataset_path)
        fetch_json['pipeline'][0]['filename'] = full_dataset_path
        fetch_json['pipeline'][0]['bounds'] = bound
        fetch_json['pipeline'][1]['polygon'] = polygon_input
        fetch_json['pipeline'][5]['filename']=f"data/{region}.laz"
        fetch_json['pipeline'][6]['filename']=f"data/{region}.tif"
        pipeline = pdal.Pipeline(json.dumps(fetch_json))
        return pipeline

    def runPipeline(self, polygon: Polygon):
        pipeline = self.getPipeline(polygon)
        try:
            pipeline.execute()
            metadata = pipeline.metadata
            log = pipeline.log
            return pipeline.arrays
        except RuntimeError as e:
            print(e)
    def get_elevation(self, array_data):
        if array_data:

            for i in array_data:
                geometry_points = [Point(x, y) for x, y in zip(i["X"], i["Y"])]
                elevations = i["Z"]
                df = gpd.GeoDataFrame(columns=["elevation", "geometry"])
                df['elevation'] = elevations
                df['geometry'] = geometry_points
                df = df.set_geometry("geometry")
                df.set_crs(epsg=26915, inplace=True)

            return df
    def visualize (self, df: gpd.GeoDataFrame, cmap="terrain") -> None:

        fig, ax = plt.subplots(1, 1, figsize=(12, 10))

        df.plot(column='elevation', ax=ax, legend=True, cmap=cmap)
        plt.show()
if __name__ == "__main__":

    MINX, MINY, MAXX, MAXY = [-93.756155, 41.918015, -93.747334, 41.921429]

    polygon = Polygon(((MINX, MINY), (MINX, MAXY),
                       (MAXX, MAXY), (MAXX, MINY), (MINX, MINY)))
    bound="([-10425171.940, -10423171.940], [5164494.710, 5166494.710])"
    region="IA_FullState"
    fetcher = Fetch_data(bound,region)
    data=fetcher.runPipeline(polygon)
#     df = fetcher.get_elevation(data)
#     print(df.info())
#     print(df)
